### PR TITLE
Feature request: Invites revamp

### PR DESCRIPTION
**Pull request by aligatorr89**
Feature request: https://taggr.network/#/post/1333906

- Extend invites to struct: credits, credits_per_user, realm_id, joined_by_user_ids
- Change create_invite
   - add new parameters and validate edge cases
- Update update_invite:
  -  owner can increase credits allowed to consume on invite
  -   owner can change realm_id
- Implement frontend requirements
  - Revamp invite page
  - Registration flow: if invite has realm_id, navigate user to that realm. Avoid showing him Mint credits page
- Protect transfering credits to 1 account (any of below it passes) `user::validate_send_credits`
  - invited user consumed amount of credits from invite
  - invited user purchased new credits
- Addtionally: create_invite has small vulnerability, anyone can create Infinite number of invites without paying anything. It can consume heap memory. Fixed within `invite::validate_user_invites_credits`
  - checking if ALL invites credits are smaller than owners amount of credits

### CODE REVIEW

**Comment by aligatorr89** [comment]
Had small issues because timer is mocked to 0 in tests. I had to revert some new tests due to this issue

**Comment by xqxpx** [comment]
Will review soon! 👍

File: src/backend/env/invite.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+use std::collections::BTreeSet;
+
+#[derive(Default, Serialize, Deserialize, Clone)]
+pub struct Invite {
+    pub credits: Credits,
+    pub credits_per_user: Credits,
+    pub joined_user_ids: BTreeSet<UserId>,
+    pub realm_id: Option<RealmId>,
+    /// Owner id
+    pub user_id: UserId,
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Should we use a name so that 1) the comment is not needed and 2) the semantics is clear on any line where it is used, e.g. `inviter`?

File: src/backend/env/invite.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+impl Invite {
+    pub fn new(
+        credits: Credits,
+        credits_per_user: Credits,
+        realm_id: Option<RealmId>,
+        user_id: UserId,
+    ) -> Self {
+        // Convert empty realm_id to None
+        let converted_relm_id: Option<RealmId> = realm_id.filter(|id| !id.is_empty());
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Is this the right way, though? Shouldn't we instead check that the realm exist at all? 

If there is no easy way to do it here, alternatively, can we handle the case later? E.g. if the realm_id is present, but no realm can be found, we default to the None case?

I just see no good reason for handling one special case for `""` while it still can be `"abra-cadabra"` :)

File: src/backend/env/invite.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    pub fn consume(&mut self, joined_user_id: UserId) -> Result<(), String> {
+        let new_credits = self
+            .credits
+            .checked_sub(self.credits_per_user)
+            .ok_or("Invite credits too low")?;
+        self.credits = new_credits;
+
+        if self.joined_user_ids.contains(&joined_user_id) {
+            return Err("User already credited".into());
+        }
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Please move this to the top of the function. If we'd exit with an error here, we leave the state mutated (credits subtracted).

File: src/backend/env/invite.rs [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            self.credits = new_credits;
+        }
+
+        if let Some(id) = realm_id {
+            if id.is_empty() {
+                self.realm_id = None;
+            } else {
+                self.realm_id = Some(id);
+            }
+        }
```
Comment (taggrx) [inline src/backend/env/invite.rs]: WDYT, how real is this case? To me it feels like a 99.99% edge case, but I might be wrong. In case you agree: should we maybe simplify and not allow it at all? If somebody would need it, they can create a new invite.

Just in case you agree: I'd also rename the function to `top_up`.

File: src/backend/env/invite.rs [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            return Err("Invite credits overflow".into());
+        }
+    }
+
+    Ok(())
+}
+
+pub fn validate_realm_id(state: &State, realm_id: Option<&RealmId>) -> Result<(), String> {
+    if let Some(id) = realm_id {
+        if !id.is_empty() && !state.realms.contains_key(id) {
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Wouldn't `state.realms.contains_key("")` return `false` anyway?

File: src/backend/env/invite.rs [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+pub fn invites_by_principal(state: &State, principal: Principal) -> Vec<(String, Invite)> {
+    state
+        .principal_to_user(principal)
+        .map(|user| {
+            state
+                .invite_codes
+                .iter()
+                .filter(|(_, invite)| invite.user_id == user.id)
+                .map(|(code, invite)| (code.clone(), invite.clone()))
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Is there a good reason for cloning instead of returning references? If you struggle with life-times, let me know, I can rewrite this part.

File: src/backend/env/invite.rs [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                .collect::<Vec<_>>()
+        })
+        .unwrap_or_default()
+}
+
+pub fn validate_user_invites_credits(
```
Comment (taggrx) [inline src/backend/env/invite.rs:100]: I (and I hope many) really would appreciate having explanatory docs on what this is doing, before digging into code.

File: src/backend/env/invite.rs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+            if total_with_diff > user.credits() {
+                return Err(format!(
+                    "You don't have enough credits to support invites {} , {}",
+                    user.credits(),
+                    total_with_diff
+                ));
+            }
+        }
+        None => {
+            return Err("Invite credits overflow".into());
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Can't we just handle it with `ok_or` in line 110?

File: src/backend/env/mod.rs [hunk 9]
```diff
@@ -989,26 +1004,79 @@ impl State {
...
+    pub fn create_invite(
+        &mut self,
+        principal: Principal,
+        credits: Credits,
+        credits_per_user_opt: Option<Credits>,
+        realm_id: Option<RealmId>,
+    ) -> Result<(), String> {
+        let credits_per_user = credits_per_user_opt.unwrap_or(credits);
+        if credits % credits_per_user != 0 {
+            return Err("Credits per user are not a multiplier of credits".into());
```
Comment (taggrx) [inline src/backend/env/mod.rs]:
```suggestion
            return Err("Credits per user are not a multiple of credits".into());
```

File: src/backend/env/mod.rs [hunk 10]
```diff
@@ -989,26 +1004,79 @@ impl State {
...
             return Err(format!(
                 "smallest invite must contain {} credits",
                 min_credits
             ));
         }
-        if user.credits() < credits {
-            return Err("not enough credits".into());
-        }
+
+        invite::validate_realm_id(self, realm_id.as_ref())?;
```
Comment (taggrx) [inline src/backend/env/mod.rs]: I think realm validation is the concern of the state module, so I think `validate_realm_id` could be just an private function on `state` itself.

File: src/backend/env/mod.rs [hunk 11]
```diff
@@ -989,26 +1004,79 @@ impl State {
...
+            .ok_or(format!("Invite '{}' not found", invite_code))?;
+        if let Some(credits) = credits {
+            invite::validate_user_invites_credits(self, user, credits, Some(*invite_credits))?;
+
+            if credits % credits_per_user != 0 {
+                return Err(format!(
+                    "Credits per user are not a multiplier of credits {} {}",
+                    credits, credits_per_user
+                ));
+            }
```
Comment (taggrx) [inline src/backend/env/mod.rs]: Shouldn't this be part of `validate_user_invites_credits`?

File: src/backend/env/mod.rs [hunk 12]
```diff
@@ -989,26 +1004,79 @@ impl State {
...
+        }
+
+        self.invite_codes
+            .get_mut(&invite_code)
+            .ok_or(format!("Invite '{}' not found", invite_code))?
+            .update(credits, realm_id, user_id)?;
```
Comment (taggrx) [inline src/backend/env/mod.rs:1083]: Would it make sense for `update` to run `validate_user_invites_credits` first? So that the caller doesn't need to think about validation? (this sounds more robust to me)

File: src/backend/env/user.rs [hunk 13]
```diff
@@ -611,6 +611,21 @@ impl User {
...
         })
     }
+
+    pub fn validate_send_credits(&self, state: &State) -> Result<(), String> {
+        if let Some(invited_by) = self.invited_by {
+            let invite = state.invite_codes.values().find(|invite| {
+                invited_by == invite.user_id && invite.joined_user_ids.contains(&self.id)
+            });
+            if let Some(invite) = invite {
+                if self.credits_burned() < invite.credits_per_user && self.credits() < 1000 {
```
Comment (taggrx) [inline src/backend/env/user.rs]: The first condition is an excellent idea!!!

But why do we need the check `< 1000`?

Also, what's the life cycle of an invite? I think the old ones get deleted once used.

File: src/backend/env/invite.rs (hunk 1 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Yes, I like foreign key to be the same. Since invite has 2 relations to user, I changed it to `inviter_user_id`

File: src/backend/env/invite.rs (hunk 3 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Moved to the top. It might be better to panic which reverts state changes under mutate.

File: src/backend/env/user.rs (hunk 13 above)
Comment (aligatorr89) [reply src/backend/env/user.rs]: Removed it. It was supposed to allow users that purchased credits, but attacker can simply send 1000 to all accounts and send 1050 back. 

Invite doesn't get removed anymore, it can be topped up when it reaches 0 credits. Use case will be:

- print qr code and paste on university/town-hall wall
- top it up based on how promotion is going

File: src/backend/env/invite.rs (hunk 8 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Yes, I've simplified it .

File: src/backend/env/mod.rs (hunk 10 above)
Comment (aligatorr89) [reply src/backend/env/mod.rs]: Do you think any function that uses state directly should be in mod.rs ?

File: src/backend/env/mod.rs (hunk 12 above)
Comment (aligatorr89) [reply src/backend/env/mod.rs:1083]: validate_user_invites_credits uses state, while invite update only uses self (invite instance)

What should be done here in general? I've seen different approaches in codebase.

File: src/backend/env/invite.rs (hunk 5 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: idea is that "" is allowed as a way to unset to None

File: src/backend/env/invite.rs (hunk 3 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Tbh, I was under the impression that mutate will auto-revert changes if result is err too. Thanks for spotting this!

File: src/backend/env/invite.rs (hunk 2 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: It's done in create_invite in mod.rs . The reason is it uses global state.

File: src/backend/env/invite.rs (hunk 4 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: I think it's nice to have. QR codes will be in phisical locations, so it's nice to have ability to change realm on them.

File: src/backend/env/mod.rs (hunk 11 above)
Comment (aligatorr89) [reply src/backend/env/mod.rs]: validate_user_invites_credits is based on all user invites in state. I moved it to invite::update

File: src/backend/env/invite.rs (hunk 3 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: No, this in impossible to do in a general case, without cloning the whole state right?

File: src/backend/env/mod.rs (hunk 10 above)
Comment (taggrx) [reply src/backend/env/mod.rs]: Definitely not any, but realm validation sound like a useful function and can be probably used somewhere else too?

File: src/backend/env/invite.rs (hunk 5 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: Hm, sorry I still don't understand, please explain with an example?

File: src/backend/env/invite.rs (hunk 2 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: I think I do not understand the `Some("")` case at all yes, but I asked a related question on other thread.

File: src/backend/env/invite.rs (hunk 4 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: When is such an invite expected to be clean up?

File: src/backend/env/mod.rs (hunk 12 above)
Comment (taggrx) [reply src/backend/env/mod.rs:1083]: Oh I see the problem. Yeah, this is an ugly artifact of Rust... No, then let it as is.

File: src/backend/env/mod.rs [hunk 14]
```diff
@@ -4843,6 +4912,46 @@ pub(crate) mod tests {
...
+            code
+        });
+
+        // New user should be joined to realm
+        let new_principal = pr(5);
+        assert_eq!(
+            State::create_user(new_principal, "name".to_string(), Some(invite_code)).await,
+            Ok(Some(realm_id.clone()))
+        );
+        read(|state| {
```
Comment (taggrx) [inline src/backend/env/mod.rs:4945]: Should we also assert that user has a credit balance we expect and the invite's balance decreased as expected?

File: src/backend/env/invite.rs (hunk 4 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: It will never clean up. We have validation against creating invites over balance, so it won't be infinite.

 We could cleanup invites that run out of credits and were not used in long time?

File: src/backend/env/invite.rs (hunk 5 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: when you call update, you can unset it by "". Because of Option<> input, None means to ignore, while "" means to unset.

File: src/backend/env/mod.rs (hunk 14 above)
Comment (aligatorr89) [reply src/backend/env/mod.rs:4945]: ofc, added

File: src/backend/env/mod.rs (hunk 10 above)
Comment (aligatorr89) [reply src/backend/env/mod.rs]: Yes, i moved it to mod.

File: src/backend/env/invite.rs (hunk 6 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Will refactor to Box, which is fixed size and more performant than Vec

File: src/backend/env/invite.rs (hunk 4 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: Ok, maybe it's not urgent right now.

File: src/backend/env/invite.rs (hunk 5 above)
Comment (taggrx) [reply src/backend/env/invite.rs]: Correct, but we will never have `state.realms.contains_key(id)` being true in this case, right? So why the additional check for emptyness?

Comment (aligatorr89) [reply src/backend/env/invite.rs]: Yes, but we want "" to skip check as it will be converted to None

Comment (aligatorr89) [reply src/backend/env/invite.rs]: ah, I see, sec... :D

Comment (aligatorr89) [reply src/backend/env/invite.rs]: Basically, we want "" to pass, cause it will be converted to None

File: src/backend/env/invite.rs [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+impl Invite {
+    pub fn new(
+        credits: Credits,
+        credits_per_user: Credits,
+        realm_id: Option<RealmId>,
+        inviter_user_id: UserId,
+    ) -> Self {
+        // Convert empty realm_id to None
+        let converted_relm_id: Option<RealmId> = realm_id.filter(|id| !id.is_empty());
```
Comment (taggrx) [inline src/backend/env/invite.rs]:
```suggestion
        let converted_realm_id: Option<RealmId> = realm_id.filter(|id| !id.is_empty());
```

File: src/backend/env/invite.rs [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+
+        Self {
+            credits,
+            credits_per_user,
+            joined_user_ids: BTreeSet::new(),
+            realm_id: converted_relm_id,
```
Comment (taggrx) [inline src/backend/env/invite.rs]:
```suggestion
            realm_id: converted_realm_id,
```

File: src/backend/env/invite.rs [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    ) -> Result<(), String> {
+        if self.inviter_user_id != user_id {
+            return Err("Owner does not match".into());
+        }
+
+        if let Some(new_credits) = credits {
+            if new_credits % self.credits_per_user != 0 {
+                return Err(format!(
+                    "Credits per user are not a multiple of credits {} {}",
+                    new_credits, self.credits_per_user
```
Comment (taggrx) [inline src/backend/env/invite.rs]:
```suggestion
                    "Credits per user {} are not a multiple of new credits {}",
                    self.credits_per_user, new_credits,
```

File: src/backend/env/invite.rs [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+        .map(|user_id| {
+            state
+                .invite_codes
+                .iter()
+                .filter(move |(_, invite)| invite.inviter_user_id == user_id)
+        })
+        .expect("Failed to filter invites");
+
+    Box::new(invites)
+}
```
Comment (taggrx) [inline src/backend/env/invite.rs]: How about avoiding panicking?

```suggestion
pub fn invites_by_principal(
    state: &State,
    principal: Principal,
) -> Box<dyn Iterator<Item = (&'_ String, &'_ Invite)> + '_> {
    match state.principal_to_user(principal).map(|user| {
        state
            .invite_codes
            .iter()
            .filter(move |(_, invite)| invite.inviter_user_id == user.id)
    }) {
        Some(iter) => Box::new(iter),
        _ => Box::new(std::iter::empty()),
    }
}
```

File: src/backend/env/invite.rs [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+/**
+ * Check allocated credits in invites do not exceed user's credits balance
+ *
+ * Protects against creating infinite number of invites
+ */
```
Comment (taggrx) [inline src/backend/env/invite.rs]: Let's stick with the [RustDoc](https://doc.rust-lang.org/rustdoc/what-is-rustdoc.html) style?

```suggestion
/// Check allocated credits in invites do not exceed user's credits balance.
/// Protects against creating infinite number of invites.
```

File: src/backend/env/mod.rs [hunk 20]
```diff
@@ -996,26 +1011,71 @@ impl State {
...
+        invite_code: String,
+        credits: Option<Credits>,
+        realm_id: Option<RealmId>,
+    ) -> Result<(), String> {
+        if credits.is_none() && realm_id.is_none() {
+            return Err("Update is empty".into());
+        }
+        let Some(user) = self.users.get(&user_id) else {
+            return Err("User not found".into());
+        };
```
Comment (taggrx) [inline src/backend/env/mod.rs]:
```suggestion
        let user = self.users.get(&user_id).ok_or("user not found")?;
```

File: src/backend/env/invite.rs [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+        .checked_add(new_credits)
+        .ok_or("Invite credits overflow")?
+        .checked_sub(old_credits.unwrap_or_default())
+        .ok_or("Invite credits overflow")?;
+
+    if total_with_diff > user.credits() {
+        return Err(format!(
+            "You don't have enough credits to support invites {} , {}",
+            user.credits(),
+            total_with_diff
```
Comment (taggrx) [inline src/backend/env/invite.rs]:
```suggestion
            "not enough credits available: {} (needed for all open invites: {})",
            user.credits(),
            total_with_diff
```

File: src/backend/env/mod.rs [hunk 22]
```diff
@@ -910,29 +916,50 @@ impl State {
...
+            if let Some((credits_per_user, inviter_id, code, realm_id)) =
+                invite_code.and_then(|code| {
+                    let invite = state.invite_codes.get(&code)?;
+                    Some((
+                        invite.credits_per_user,
+                        invite.inviter_user_id,
+                        code,
+                        invite.realm_id.clone(),
+                    ))
+                })
```
Comment (taggrx) [inline src/backend/env/mod.rs]:
```suggestion
            if let Some((credits_per_user, inviter_id, code, realm_id)) =
                invite_code.and_then(|code| {
                    state.invite_codes.get(&code).map(|invite| {
                        (
                            invite.credits_per_user,
                            invite.inviter_user_id,
                            code,
                            invite.realm_id.clone(),
                        )
                    })
                })
```

File: src/backend/updates.rs [hunk 23]
```diff
@@ -332,8 +305,18 @@ fn mint_credits() {
...
+    let (credits, credits_per_user, realm_id): (Credits, Option<Credits>, Option<RealmId>) =
+        parse(&arg_data_raw());
+    mutate(|state| reply(state.create_invite(caller(), credits, credits_per_user, realm_id)));
+}
+
+#[export_name = "canister_update update_invite"]
+fn update_invite() {
+    let (invite_code, credits, realm_id): (String, Option<Credits>, Option<RealmId>) =
+        parse(&arg_data_raw());
+    let user_id = read(|state| state.principal_to_user(caller()).expect("no user found").id);
```
Comment (taggrx) [inline src/backend/updates.rs]: If I remember correctly this kind of app data is always extracted inside modules, not in `updates.rs` which mostly contains "plumbing" code.

Is there any reason to not simply pass `caller()` as a first argument and then find user and do all error handling inside `update_invite`?

File: src/frontend/src/invites.tsx [hunk 24]
```diff
@@ -17,6 +35,56 @@ export const Invites = () => {
...
+            return;
+        }
+        return window.api
+            .call<any>(
+                "update_invite",
+                code,
+                updatedInvite.credits !== undefined &&
+                    updatedInvite.credits >= 0
+                    ? updatedInvite.credits
+                    : null,
```
Comment (taggrx) [inline src/frontend/src/invites.tsx]: Wouldn't this work too here?
```suggestion
                updatedInvite.credits || null
```

File: src/frontend/src/invites.tsx [hunk 25]
```diff
@@ -17,6 +35,56 @@ export const Invites = () => {
...
+                updatedInvite.realm_id !== undefined
+                    ? updatedInvite.realm_id
+                    : null,
```
Comment (taggrx) [inline src/frontend/src/invites.tsx]: Same as above?

File: src/frontend/src/invites.tsx [hunk 26]
```diff
@@ -17,6 +35,56 @@ export const Invites = () => {
...
+            )
+            .then(async (response) => {
+                if ("Err" in (response || {})) {
+                    alert(`Error: ${response.Err}`);
+                    setBusy(true);
+                    await loadInvites(); // Set back to prior state
+                    setBusy(false);
```
Comment (taggrx) [inline src/frontend/src/invites.tsx]: Why do we need this? The state in UI should not have changed during the call, right?

File: src/frontend/src/invites.tsx [hunk 27]
```diff
@@ -53,6 +124,22 @@ export const Invites = () => {
...
+                        }
+                    />
+                    Realm:
+                    <input
+                        type="text"
+                        value={inviteRealm}
+                        className="max_width_col"
+                        onChange={(event) => setInviteRealm(event.target.value)}
+                    />
                     {!busy && (
```
Comment (taggrx) [inline src/frontend/src/invites.tsx]: I don't think we need this. Check out `ButtonWithLoading` instead. It simply takes an async function and displays loading while the function is called.

**Review by xqxpx — COMMENTED** [review]
Please don't forget to lowercase all error messages 🙏

File: src/frontend/src/invites.tsx (hunk 26 above)
Comment (aligatorr89) [reply src/frontend/src/invites.tsx]: This is to reset to previous state if update fails.

File: src/frontend/src/invites.tsx (hunk 24 above)
Comment (aligatorr89) [reply src/frontend/src/invites.tsx]: 0 is valid number while null means empty.

File: src/backend/updates.rs (hunk 23 above)
Comment (aligatorr89) [reply src/backend/updates.rs]: Fully agreed, I changed it.

File: src/backend/env/invite.rs (hunk 18 above)
Comment (aligatorr89) [reply src/backend/env/invite.rs]: Much better, thanks!

**Comment by aligatorr89** [comment]
> Please don't forget to lowercase all error messages 🙏

Done

**Comment by xqxpx** [comment]
@aligatorr89 I added in 1332eb803e3d2b4b0897672bbf385404068e021e a quick test for the invite validation and instantly noticed a few things that could be improved.

Could you please also add a few tests for `consume` and `update` checking the edge cases?

File: src/backend/env/user.rs [hunk 28]
```diff
@@ -611,6 +611,22 @@ impl User {
             Ok(())
         })
     }
+
+    /** Protect against invite phishing */
```
Comment (taggrx) [inline src/backend/env/user.rs]: Rustdoc please 😅

File: src/backend/env/user.rs [hunk 29]
```diff
@@ -611,6 +611,22 @@ impl User {
...
+    pub fn validate_send_credits(&self, state: &State) -> Result<(), String> {
```
Comment (taggrx) [inline src/backend/env/user.rs:618]: We should add a todo here: since credits burned get reset every week, some users will never be able to transfer credits until the invite is alive. But we can handle this at some point later.

**Review by xqxpx — APPROVED** [review]

File: src/frontend/src/invites.tsx [hunk 30]
```diff
@@ -17,95 +37,229 @@ export const Invites = () => {
...
+                alert(`Error: ${response.Err}`);
+                setBusy(true);
+                await loadInvites(); // Set back to prior state
+                setBusy(false);
+                return;
+            }
+        }
+    };
+
+    const updateInvite = (
```
Comment (taggrx) [inline src/frontend/src/invites.tsx:76]: @aligatorr89 I changed the way invites are updated (modify in-place and set the `dirty` flag).

This is because I uncovered a bug: whenever I would update credits, the realm id would be deleted because the previous code would not send the realm id as updated value.

Comment (aligatorr89) [reply src/frontend/src/invites.tsx:76]: Ok, I noticed, and added to tests

File: src/backend/env/user.rs (hunk 29 above)
Comment (aligatorr89) [reply src/backend/env/user.rs:618]: Quite a surprise we don't keep the number. We could add a metric `credits_burned_total` which wouldn't reset?
